### PR TITLE
Validate type hints with mypy, and use mypy in our CI

### PR DESCRIPTION
**Pull request by muddyfish**
## Description

Fix our type hinting, and adds mypy checks to our CI.
Update our DEVELOPMENT.md to add instructions for how to run mypy locally.

Mypy is only ran on source files right now, not tests as they need more work to pass.

I'm ignoring one mypy failure in `S3Writer`'s `write` method, as fixing it would require a larger amount of work. In particular, `BufferedIOBase`'s `write` method takes a `Buffer`, which isn't supported right now. 

## Additional context

Changes the constructor of `S3Writer.__init__` to not use `None` as a default argument for `get_object_info` and `get_stream`, as these weren't valid values. This is fine, as `S3Writer`'s constructor is considered private. 

There is not any performance impact with these changes

## Testing

https://github.com/awslabs/s3-connector-for-pytorch/actions/runs/8143889830/job/22256748400 runs and passes mypy.

--------
By submitting this pull request, I confirm that my contribution is made under the terms of BSD 3-Clause License and I agree to the terms of the [LICENSE](https://github.com/awslabs/s3-connector-for-pytorch/blob/main/LICENSE).

### CODE REVIEW

File: .github/workflows/python-checks.yml [hunk 1]
```diff
@@ -88,18 +88,25 @@ jobs:
       - name: Install Python dependencies
```
Comment (dnanuti) [inline .github/workflows/python-checks.yml:88]: Nit: please update the PR title

File: s3torchconnectorclient/MANIFEST.in [hunk 2]
```diff
@@ -1,3 +1,4 @@
 # MANIFEST.in
 include Cargo.toml
 recursive-include rust *.rs
+global-include py.typed
```
Comment (dnanuti) [inline s3torchconnectorclient/MANIFEST.in:4]: nit: new line

File: s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi [hunk 3]
```diff
@@ -112,3 +118,5 @@ class ListObjectStream:
 
 class S3Exception(Exception):
     pass
+
+__version__: str
```
Comment (dnanuti) [inline s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi:122]: Why is this here?

File: .github/workflows/python-checks.yml (hunk 1 above)
Comment (dnanuti) [reply .github/workflows/python-checks.yml:88]: Also worth mentioning that this is used just for type hinting and does not have any runtime performance implications.

**Review by dnanuti — APPROVED** [review]

File: .github/workflows/python-checks.yml (hunk 1 above)
Comment (muddyfish) [reply .github/workflows/python-checks.yml:88]: Theoretically it could have a very slight performance impact with the `assert ... is not None` if Python's ran in debug mode. I don't think it's worth worrying about

File: s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi (hunk 3 above)
Comment (muddyfish) [reply s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi:122]: We export `__version__` from Rust.

File: s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi [hunk 4]
```diff
@@ -6,23 +6,23 @@ from typing import List, Optional
...
         self,
         region: str,
-        endpoint: str = "",
         user_agent_prefix: str = "",
         throughput_target_gbps: float = 10.0,
         part_size: int = 8 * 1024 * 1024,
-        profile: str = None,
+        profile: Optional[str] = None,
         no_sign_request: bool = False,
+        endpoint: Optional[str] = None,
```
Comment (IsaevIlya) [inline s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi:25]: Why do we need to rearrange parameters order here?

File: .github/workflows/python-checks.yml (hunk 1 above)
Comment (dnanuti) [reply .github/workflows/python-checks.yml:88]: I meant the type hinting itself, not the assertions.

Comment (dnanuti) [reply .github/workflows/python-checks.yml:88]: But I agree those are negligible.

File: s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi (hunk 4 above)
Comment (muddyfish) [reply s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi:25]: They were in the wrong order - in [MountpointS3Client](https://github.com/awslabs/s3-connector-for-pytorch/blob/main/s3torchconnectorclient/rust/src/mountpoint_s3_client.rs#L56), they're written in the new order.

**Review by IsaevIlya — APPROVED** [review]
